### PR TITLE
[HLSL] implement `mad` intrinsic

### LINKED ISSUE #83736: [HLSL] implement `mad` intrinsic
https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/mad

This intrinsic is used by integer dot product and blocks completion of [#83626](https://github.com/llvm/llvm-project/issues/83626)

### PR DESCRIPTION
**Pull request by farzonl**
This change implements #83736
The dot product lowering needs a tertiary multipy add operation. DXIL has three mad opcodes for `fmad`(46), `imad`(48), and `umad`(49). Dot product in DXIL only uses `imad`\ `umad`, but for completeness and because the hlsl `mad` intrinsic requires it `fmad` was also included. Two new intrinsics were needed to be created to complete this change. the `fmad` case already supported by llvm via `fmuladd` intrinsic.

- `hlsl_intrinsics.h` - exposed mad api call.
- `Builtins.td` - exposed a `mad` builtin.
- `Sema.h` - make `tertiary` calls check for float types optional. 
- `CGBuiltin.cpp` - pick the intrinsic for singed\unsigned & float also reuse `int_fmuladd`. 
- `SemaChecking.cpp` - type checks for `__builtin_hlsl_mad`. 
- `IntrinsicsDirectX.td` create the two new intrinsics for `imad`\`umad`/ 
- `DXIL.td` - create the llvm intrinsic to  `DXIL` opcode mapping.

### CODE REVIEW

File: llvm/include/llvm/IR/IntrinsicsDirectX.td [hunk 1]
```diff
@@ -31,4 +31,8 @@ def int_dx_lerp :
     Intrinsic<[LLVMScalarOrSameVectorWidth<0, LLVMVectorElementType<0>>],
     [llvm_anyvector_ty, LLVMScalarOrSameVectorWidth<0, LLVMVectorElementType<0>>,LLVMScalarOrSameVectorWidth<0, LLVMVectorElementType<0>>],
     [IntrNoMem, IntrWillReturn] >;
+
+
```
Comment (python3kgae) [inline llvm/include/llvm/IR/IntrinsicsDirectX.td:35]:
```suggestion

```

**Review by python3kgae — APPROVED** [review]

File: llvm/test/CodeGen/DirectX/fmad.ll [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+; CHECK:call half @dx.op.tertiary.f16(i32 46, half %{{.*}}, half %{{.*}}, half %{{.*}})
+; CHECK:call float @dx.op.tertiary.f32(i32 46, float %{{.*}}, float %{{.*}}, float %{{.*}})
+; CHECK:call double @dx.op.tertiary.f64(i32 46, double %{{.*}}, double %{{.*}}, double %{{.*}})
+
+
+target datalayout = "e-m:e-p:32:32-i1:32-i8:8-i16:16-i32:32-i64:64-f16:16-f32:32-f64:64-n8:16:32:64"
+target triple = "dxil-pc-shadermodel6.7-library"
+
+; Function Attrs: noinline nounwind optnone
+define noundef half @"?test_mad_half@@YA$f16@$f16@00@Z"(half noundef %p0, half noundef %p1, half noundef %p2) #0 {
```
Comment (bharadwajy) [inline llvm/test/CodeGen/DirectX/fmad.ll:13]: Consider using unmangled names for function names in this as well as other test sources of this PR.

**Review by bharadwajy — APPROVED** [review]

File: clang/lib/Sema/SemaChecking.cpp [hunk 3]
```diff
@@ -19800,7 +19808,8 @@ bool Sema::SemaBuiltinVectorMath(CallExpr *TheCall, QualType &Res) {
   return false;
 }
 
-bool Sema::SemaBuiltinElementwiseTernaryMath(CallExpr *TheCall) {
+bool Sema::SemaBuiltinElementwiseTernaryMath(CallExpr *TheCall,
+                                             bool enforceFloatingPointCheck) {
```
Comment (bogner) [inline clang/lib/Sema/SemaChecking.cpp:19812]: Formal arguments are capitalized in LLVM's style, like "EnforceFloatingPointCheck". I'm also not a huge fan of this name - it isn't obvious what it means without looking at the implementation of the function.

File: clang/lib/Sema/SemaChecking.cpp [hunk 4]
```diff
@@ -19812,11 +19821,13 @@ bool Sema::SemaBuiltinElementwiseTernaryMath(CallExpr *TheCall) {
...
-  int ArgOrdinal = 1;
-  for (Expr *Arg : Args) {
-    if (checkFPMathBuiltinElementType(*this, Arg->getBeginLoc(), Arg->getType(),
-                                      ArgOrdinal++))
-      return true;
+  if (enforceFloatingPointCheck) {
+    int ArgOrdinal = 1;
+    for (Expr *Arg : Args) {
+      if (checkFPMathBuiltinElementType(*this, Arg->getBeginLoc(),
+                                        Arg->getType(), ArgOrdinal++))
```
Comment (bogner) [inline clang/lib/Sema/SemaChecking.cpp:19828]: Should this call some other element type check instead in this case? I initially thought `checkMathBuiltinElementType` would make sense, but from a quick glance that seems to be tied to matrix types for some reason.

File: clang/lib/Sema/SemaChecking.cpp [hunk 5]
```diff
@@ -5298,6 +5298,14 @@ bool Sema::CheckHLSLBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
       return true;
     break;
   }
+  case Builtin::BI__builtin_hlsl_mad: {
+    if (checkArgCount(*this, TheCall, 3))
+      return true;
+    if (CheckVectorElementCallArgs(this, TheCall))
+      return true;
+    if (SemaBuiltinElementwiseTernaryMath(TheCall, false))
```
[the remark this answers is not in the document]
Comment (farzonl) [reply clang/lib/Sema/SemaChecking.cpp:5306]: i usually do, but this one shows up via vscode clangd extension and it threw me off.

File: clang/lib/Sema/SemaChecking.cpp (hunk 4 above)
Comment (farzonl) [reply clang/lib/Sema/SemaChecking.cpp:19828]: mad is suppose to operate on matrix types, we just don't have them yet.

Comment (bogner) [reply clang/lib/Sema/SemaChecking.cpp:19828]: I was mostly commenting on the jump from "math builtin" to "matrix type" as far as naming goes, but I don't spend a lot of time in Sema. My point is that I think we may still want some check here, I'm just not sure what without looking at it in more detail.

File: clang/lib/Sema/SemaChecking.cpp (hunk 3 above)
Comment (farzonl) [reply clang/lib/Sema/SemaChecking.cpp:19812]: renamed to `CheckForFloatArgs`

File: clang/lib/Sema/SemaChecking.cpp (hunk 4 above)
Comment (farzonl) [reply clang/lib/Sema/SemaChecking.cpp:19828]: So looking at `checkMathBuiltinElementType` it use of `isValidElementType` works for more than just matrix types  so  i think it is fine to use.  SO to be clear below is what you would like to see?

```cpp
if (CheckForFloatArgs) {
    int ArgOrdinal = 1;
    for (Expr *Arg : Args) {
      if (checkFPMathBuiltinElementType(*this, Arg->getBeginLoc(),
                                        Arg->getType(), ArgOrdinal++))
        return true;
    }
  } else  {
    int ArgOrdinal = 1;
    for (Expr *Arg : Args) {
      if (checkMathBuiltinElementType(*this, Arg->getBeginLoc(),
                                        Arg->getType(), ArgOrdinal++))
        return true;
}
```

Comment (farzonl) [reply clang/lib/Sema/SemaChecking.cpp:19828]: The impact is it just changed the error when we use bool

Before change:
`// expected-error@-1 {{arguments are of different types ('double' vs 'bool')}}`
After change:
`// expected-error@-1 {{3rd argument must be a vector, integer or floating point type (was 'bool')}}`

**Review by bogner — APPROVED** [review]

**Pull request merged**